File: src/algorithms/python/union-find.py
# Union-Find (Disjoint Set) with Path Compression and Union by Rank

class UnionFind:
    def __init__(self, size):
        self.parent = [i for i in range(size)]
        self.rank = [0] * size

    def find(self, i):
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def union(self, i, j):
        root_i = self.find(i)
        root_j = self.find(j)
        
        if root_i != root_j:
            if self.rank[root_i] < self.rank[root_j]:
                self.parent[root_i] = root_j
            elif self.rank[root_i] > self.rank[root_j]:
                self.parent[root_j] = root_i
            else:
                self.parent[root_j] = root_i
                self.rank[root_i] += 1

arr = [0,1, 1,2, 3,4, 4,5, 2,5] # pairs to union
uf = UnionFind(6)

for i in range(0, len(arr), 2):
    uf.union(arr[i], arr[i+1])

File: src/algorithms/cpp/union-find.cpp
// Union-Find (Disjoint Set) with Path Compression and Union by Rank
#include <iostream>
#include <vector>

using namespace std;

class UnionFind {
    vector<int> parent;
    vector<int> rank;

public:
    UnionFind(int n) {
        parent.resize(n);
        rank.resize(n, 0);
        for (int i = 0; i < n; i++)
            parent[i] = i;
    }

    int find(int i) {
        if (parent[i] == i)
            return i;
        return parent[i] = find(parent[i]);
    }

    void unite(int i, int j) {
        int rootI = find(i);
        int rootJ = find(j);
        
        if (rootI != rootJ) {
            if (rank[rootI] < rank[rootJ])
                parent[rootI] = rootJ;
            else if (rank[rootI] > rank[rootJ])
                parent[rootJ] = rootI;
            else {
                parent[rootJ] = rootI;
                rank[rootI]++;
            }
        }
    }
};

int main() {
    vector<int> arr = {0,1, 1,2, 3,4, 4,5, 2,5};
    UnionFind uf(6);
    
    for (size_t i = 0; i < arr.size(); i += 2) {
        uf.unite(arr[i], arr[i+1]);
    }
    
    return 0;
}
